cmd/zed/agents/zfs_mod.c

```diff
@@ -0,0 +1,911 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2007, 2010, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012 by Delphix. All rights reserved.
+ * Copyright 2014 Nexenta Systems, Inc. All rights reserved.
+ * Copyright (c) 2016, Intel Corporation.
+ */
+
+/*
+ * ZFS syseventd module.
+ *
+ * file origin: openzfs/usr/src/cmd/syseventd/modules/zfs_mod/zfs_mod.c
+ *
+ * The purpose of this module is to identify when devices are added to the
+ * system, and appropriately online or replace the affected vdevs.
+ *
+ * When a device is added to the system:
+ *
+ * 	1. Search for any vdevs whose devid matches that of the newly added
+ *	   device.
+ *
+ * 	2. If no vdevs are found, then search for any vdevs whose udev path
+ *	   matches that of the new device.
+ *
+ *	3. If no vdevs match by either method, then ignore the event.
+ *
+ * 	4. Attempt to online the device with a flag to indicate that it should
+ *	   be unspared when resilvering completes.  If this succeeds, then the
+ *	   same device was inserted and we should continue normally.
+ *
+ *	5. If the pool does not have the 'autoreplace' property set, attempt to
+ *	   online the device again without the unspare flag, which will
+ *	   generate a FMA fault.
+ *
+ *	6. If the pool has the 'autoreplace' property set, and the matching vdev
+ *	   is a whole disk, then label the new disk and attempt a 'zpool
+ *	   replace'.
+ *
+ * The module responds to EC_DEV_ADD events.  The special ESC_ZFS_VDEV_CHECK
+ * event indicates that a device failed to open during pool load, but the
+ * autoreplace property was set.  In this case, we deferred the associated
+ * FMA fault until our module had a chance to process the autoreplace logic.
+ * If the device could not be replaced, then the second online attempt will
+ * trigger the FMA fault that we skipped earlier.
+ *
+ * ZFS on Linux porting notes:
+ *	In lieu of a thread pool, just spawn a thread on demmand.
+ *	Linux udev provides a disk insert for both the disk and the partition
```

Comment:
Eventually we're going to want some kind of thread pool for the zed so the zedlets can be run in parallel, we could use the same thread pool for this.  One expedient option would be to use the same workaround as in `zpool_do_import()`.  Before `zpool_search_import()` is called the user space taskq implementation is initialized and used to for the import.  It's then cleaned up after this one call.  We _could_ do something like that here and use it for the life of the zed.